cmd/lifecycle/builder.go

```diff
@@ -48,26 +60,160 @@ func (b *buildCmd) Args(nargs int, args []string) error {
 }
 
 func (b *buildCmd) Privileges() error {
-	// builder should never be run with privileges
-	if priv.IsPrivileged() {
+	_, privGroup, _, err := b.readData()
+	if err != nil {
+		return err
+	}
+
+	hasPrivGroups := len(privGroup.Group) > 0
+
+	// builder should never be run with privileges if there aren't any stack buildpacks
+	if !hasPrivGroups && priv.IsPrivileged() {
 		return cmd.FailErr(errors.New("refusing to run as root"), "build")
+	} else if hasPrivGroups && !priv.IsPrivileged() {
+		return cmd.FailErr(errors.New("must run as root"), "build")
 	}
 	return nil
 }
 
 func (b *buildCmd) Exec() error {
-	group, plan, err := b.readData()
+	group, privGroup, plan, err := b.readData()
+	if err != nil {
+		return err
+	}
+
+	if err := verifyBuildpackApis(group, privGroup); err != nil {
+		return err
+	}
+
+	if len(privGroup.Group) == 0 {
+		return b.build(group, plan)
+	}
+	return b.buildWithReexec(group, privGroup, plan)
+}
+
+func (ba buildArgs) buildAll(group, privGroup lifecycle.BuildpackGroup, plan lifecycle.BuildPlan, rootuid, rootgid, uid, gid int) error {
+	if len(privGroup.Group) > 0 {
+		if err := priv.RunAsEffective(rootuid, rootgid); err != nil {
+			return err
+		}
+		if err := ba.stackBuild(privGroup, plan); err != nil {
+			return err
+		}
+		if err := priv.RunAsEffective(uid, gid); err != nil {
+			return err
+		}
+	}
+
+	return ba.build(group, plan)
+}
+
+func (ba buildArgs) stackBuild(privGroup lifecycle.BuildpackGroup, plan lifecycle.BuildPlan) error {
+	stackSnapshotter, err := snapshot.NewKanikoSnapshotter("/", ba.layersDir, ba.platformDir)
 	if err != nil {
 		return err
 	}
-	if err := verifyBuildpackApis(group); err != nil {
+
+	workspaceDir, err := ioutil.TempDir("", "stack-workspace")
+	if err != nil {
+		return err
+	}
+
+	builder, err := ba.createBuilder(privGroup, plan)
+	if err != nil {
 		return err
 	}
-	return b.build(group, plan)
+
+	builder.Snapshotter = stackSnapshotter
+	builder.AppDir = workspaceDir
```

Comment:
This feel very backdoor to me. I don't let setting a field called `AppDir` to something that isn't the app dir. It seems like we are doing it to change as little as possible in builder but in the process we may be creating a bad interface. As a user, if I read the godoc for the `lifecycle` package it would be very hard to figure out how to run a stack build. We would need comments that said things like "To run a stack build set the AppDir to a temporary directory and supply a snapshotter." which doesn't seem like a very natural interface.

Reply:
That's fair and for sure the direction I was going (changing as little as possible). I agree that we should create a better interface, so I'll put something up around this.